akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala

```diff
@@ -0,0 +1,147 @@
+import akka.util.ByteString
+
+trait ReverseProxyDirectives {
+
+  def reverseProxy(target: ReverseProxyTargetMagnet): Route =
+    extractExecutionContext { implicit ec =>
+      extractMaterializer { implicit mat =>
+        // todo customize response
+        // drain response entity if incoming request timeout occurs
+        def processTimeout(proxyResult: Future[HttpResponse]): HttpRequest => HttpResponse = _ => {
+          proxyResult.andThen {
+            case Success(proxyResponse) => proxyResponse.discardEntityBytes()
+          }
+
+          HttpResponse(StatusCodes.ServiceUnavailable, entity =
+            HttpEntity.Strict(
+              ContentTypes.`application/json`,
+              ByteString("""{"error":"The server was not able to produce a timely response to your request."}""")
+            )
+          )
+        }
+
+        extractRequestContext { ctx =>
+          // we don't need to use request.effectiveUri here since we're going to overwrite the scheme and authority
+          val incomingUri = ctx.request.uri
+          val outgoingUri =
+            (if (target.config.useUnmatchedPath) incomingUri.withPath(ctx.unmatchedPath) else incomingUri)
+              .withAuthority(target.config.targetAuthority)
+              .withScheme(target.config.targetScheme)
+
+          val outgoingRequest = mapProxyRequest(ctx.request, target.config).withUri(outgoingUri)
+          val eventualResponse = target.httpClient(outgoingRequest)
+
+          withRequestTimeoutResponse(processTimeout(eventualResponse)) {
+            complete(eventualResponse)
+          }
+        }
+      }
+    }
+
+  // remove any headers that shouldn't be passed through a proxy
+  private def mapProxyRequest(request: HttpRequest, config: ReverseProxyTargetConfig): HttpRequest = {
+    val incomingHeaders = request.headers
+
+    val remoteAddressOption = request.header[`Remote-Address`].map(_.address)
+    val xForwardedForAddressesOption = request.header[`X-Forwarded-For`].map(_.addresses)
+
+    val updatedXRealIpHeaderOption = request.header[`X-Real-Ip`]
+      .map(_.address)
+      .orElse(xForwardedForAddressesOption.flatMap(_.headOption))
+      .orElse(remoteAddressOption)
+      .map(`X-Real-Ip`(_))
+
+    val updatedXForwardedForHeaderOption = remoteAddressOption
+      .map(_ +: xForwardedForAddressesOption.getOrElse(Nil))
+      .map(`X-Forwarded-For`(_))
+
+    val intermediary = {
+      val (protocol, version) = {
+        val parts = request.protocol.value.split('/')
+        if (parts(0) == "HTTP" && parts.length > 1) None -> parts(1)
+        else if (parts.length > 1) Some(parts(0)) -> parts(1)
+        else None -> parts(0)
+      }
+
+      config.viaId.map(ViaIntermediary(protocol, version, _))
+    }
+
+    val maybeVia = request.header[Via]
+      .map(via => via.copy(intermediaries = via.intermediaries ++ intermediary))
+      .orElse(intermediary.map(i => Via(List(i))))
+
+    val outgoingHeaders = incomingHeaders.flatMap {
```

Comment:
How was this list compiled? Is there an RFC that gives some recommendations? Should the user be able to configure it? E.g. for the ones that are just filtered-out we could give the list in the configuration where we could also document why they are filtered out.

What about headers that might carry authentication information? Should they be forwarded automatically? Or should it be opt-in?

Reply:
Hop-by-hop headers are defined in [http 1.1](https://tools.ietf.org/html/rfc2616#section-13.5.1). These don't seem to be something the user would configure. 
[7230](https://tools.ietf.org/html/rfc7230#section-5.7) talks about the via header. To me that seems more for public and/or forward proxies rather than reverse proxies. So it is something the user wants/needs to configure.
As far as a literal list, https://www.mnot.net/blog/2011/07/11/what_proxies_must_do was educational.

I think we should establish that aside from headers that are specifically relevant to proxying, we should establish the default as "the request is forwarded as-is" (which is something that should go in that documentation page you mentioned :)). This way the user doesn't form an expectation that any part of the request will be treated specially, and that they need to modify it in an outer directive if they want to change what is forwarded. For authentication specifically this prevents us from having to define what qualifies as a "default" authentication header/cookie.